File: trabalho1/kojunSolver.hs
module KojunSolver where

import Utils.Matrix

{- 
A partir de uma matriz de regiões, uma região e uma posição, obtém recursivamente todas as posições da região que contém a
posição passada como parâmetro.
-}
getRegionFromPosition :: Eq a => Matrix a -> a -> Int -> Int -> [(Int, Int)] -> [(Int, Int)]
getRegionFromPosition regionGrid region row col visited
    | row < 0 || row >= numRows regionGrid ||
      col < 0 || col >= numCols regionGrid ||
      getMatrixValue regionGrid (row, col) /= region ||
      (row, col) `elem` visited = visited
    | otherwise =
        let visited' = (row, col) : visited
        in foldr
            (\(r, c) accVisited -> getRegionFromPosition regionGrid region r c accVisited)
            visited' 
            [(row - 1, col), (row + 1, col), (row, col - 1), (row, col + 1)]

{-
Recebe uma matriz de valores (Int), uma matriz de regiões (Char), uma Position e um valor Int e verifica se o valor pode
ser posicionado em uma posição Position da matriz de valores seguindo as regras do Kojun:
1. Inserir um número em cada célula da grade para que cada região de tamanho N contenha cada número de 1 a N exatamente
uma vez.
2. Os números em células ortogonalmente adjacentes devem ser diferentes.
3. Se duas células são adjacentes verticalmente na mesma região, o número na célula superior deve ser maior que o número
na célula inferior.
-}
isValidPlacement :: Matrix Int -> Matrix Char -> Position -> Int -> Bool
isValidPlacement valueGrid regionGrid (row, col) value
        | getMatrixValue valueGrid (row, col) /= 0 = False  -- Verifica se a posição já está ocupada por algum valor diferente de 0
        | value < 1 || value > regionSize = False           -- Verifica se o número está entre 1 e N, onde N é o tamanho da região
        | hasAdjacentValue = False                          -- Verifica se algum dos valores adjacentes é igual ao valor que queremos inserir
        | hasValueInRegion = False                          -- Verifica se o valor já está na região
        | isTopInvalid = False                              -- Se a posição acima for inválida (for da mesma região e menor em valor), retorna False
        | isBottomInvalid = False                           -- Se a posição abaixo for inválida (for da mesma região e maior em valor), retorna False
        | otherwise = True
    where
        numRows' = numRows valueGrid
        numCols' = numCols valueGrid
        region = getMatrixValue regionGrid (row, col)
        
        -- Filtra as posições adjacentes que são válidas (dentro da matriz) e obtém os valores das posições adjacentes
        adjacentPositions =
            filter (\(r, c) -> r >= 0 && r < numRows' && c >= 0 && c < numCols')
            [(row - 1, col), (row + 1, col), (row, col - 1), (row, col + 1)]
        adjacentValues = map (\(r, c) -> getMatrixValue valueGrid (r, c)) adjacentPositions
        -- Obtém as posições da região que contém a posição (row, col)
        regionPositions = getRegionFromPosition regionGrid region row col []
        regionSize = length regionPositions

        -- (1) Verifica se o valor já está presente na região que contém a posição
        hasValueInRegion = value `elem` map (\(r, c) -> getMatrixValue valueGrid (r, c)) regionPositions
        -- (2) Verifica se algum dos valores adjacentes é igual ao valor que queremos inserir
        hasAdjacentValue = any (== value) adjacentValues
        -- (3) Verifica se a posição acima de (row, col) é da mesma região e se o valor é maior
        isTopSameRegion = (row - 1, col) `elem` regionPositions
        isTopInvalid = isTopSameRegion && getMatrixValue valueGrid (row - 1, col) <= value
        -- (3) Verifica se a posição abaixo de (row, col) é da mesma região e se o valor é menor
        isBottomSameRegion = (row + 1, col) `elem` regionPositions
        isBottomInvalid = isBottomSameRegion && getMatrixValue valueGrid (row + 1, col) >= value

{-
Resolve o Kojun a partir de uma matriz de valores (Int) e uma matriz de regiões (Char).
Por ser uma função recursiva, recebe também um ponto de partida (linha e coluna) para aplicar o backtracking. A função
normalmente é chamada externamente com (0, 0).
O retorno é um monad Maybe, que pode ser Nothing (caso não haja solução) ou Just Matrix Int (caso haja solução) onde a
matriz de inteiros representa a solução que foi encontrada para o Kojun.
-}
solveKojun :: Matrix Int -> Matrix Char -> Position -> Int -> Maybe (Matrix Int)
solveKojun valueGrid regionGrid (row, col) maxRegionSize
        | row == numRows valueGrid = Just valueGrid -- Caso chegou ao final da matriz
        | col == numCols valueGrid = solveKojun valueGrid regionGrid (row + 1, 0) maxRegionSize -- Caso chegou ao final da linha, pula para a próxima 
        | getMatrixValue valueGrid (row, col) /= 0 = solveKojun valueGrid regionGrid (row, col + 1) maxRegionSize -- Caso a posição já esteja ocupada, pula para a próxima
        | otherwise = tryValues valueGrid regionGrid row col [1..maxRegionSize] -- Caso nenhuma das condições acima seja satisfeita, chama tryValues com os valores de 1 a N,
    where
        {-
        Tenta colocar valores válidos na posição atual.
        Input: Recebe a matriz de valores, a matriz de regiões, a linha e coluna atual e uma lista de valores a serem testados.
        Output:
        Caso nenhum dos valores da lista seja válido (conferido por isValidPlacement), retorna Nothing.
        Caso algum valor da lista seja válido, chama recursivamente solveKojun com a matriz de valores atualizada.
        -}
        tryValues :: Matrix Int -> Matrix Char -> Int -> Int -> [Int] -> Maybe (Matrix Int)
        tryValues _ _ _ _ [] = Nothing -- Caso não há mais valores para testar, retorna Nothing
        tryValues vg rg r c (value:rest) -- Desconstrói a lista de valores em head (value) e tail (rest)
            | isValidPlacement vg rg (r, c) value =
                case solveKojun (setMatrixValue vg (r, c) value) rg (r, c + 1) maxRegionSize of -- Tenta resolver o Kojun com o valor inserido na posição atual
                    Just result -> Just result
                    Nothing -> tryValues vg rg r c rest -- Nova chamada de tryValue com o restante da lista
            | otherwise = tryValues vg rg r c rest

File: kojun-solver-scala/src/main/scala/solver/KojunSolver.scala
package solver

import utils.Matrix

import scala.annotation.tailrec
import scala.collection.mutable.ListBuffer
import scala.collection.mutable
import scala.language.postfixOps

class KojunSolver(valueGrid: Matrix[Int], regionGrid: Matrix[String]) {
  private type Position = (Int, Int)

  private val regionMap = getRegionMap

  // Gera uma hashmap onde a chave é o valor da região e o valor é uma lista de tuplas (linha, coluna)
  private def getRegionMap: mutable.HashMap[String, List[(Int, Int)]] = {
    val regionMap = new mutable.HashMap[String, ListBuffer[(Int, Int)]]()
    
    for {
      row <- 0 until regionGrid.numRows
      col <- 0 until regionGrid.numCols
    } {
      val regionValue = regionGrid.getMatrixValue((row, col))
      regionMap.getOrElseUpdate(regionValue, ListBuffer()) += ((row, col))
    }
    // Converte a hashmap para uma lista de tuplas
    val result = regionMap.map { case (regionValue, positionsBuffer) =>
      regionValue -> positionsBuffer.toList
    }

    result
  }

  /**
   * A partir da matriz de valores, matriz de regiões, hashmap de regiões, uma Position e um valor Int verifica se o
   * valor pode ser posicionado em uma posição Position da matriz de valores seguindo as regras do Kojun:
   * 1. Inserir um número em cada célula da grade para que cada região de tamanho N contenha cada número de 1 a N
   * exatamente uma vez.
   * 2. Os números em células ortogonalmente adjacentes devem ser diferentes.
   * 3. Se duas células são adjacentes verticalmente na mesma região, o número na célula superior deve ser maior que o
   * número na célula inferior.
   */
  private def canInsertValue(
    position: Position,
    value: Int,
    valueGrid: Matrix[Int],
    regionGrid: Matrix[String],
    regionMap: mutable.HashMap[String, List[(Int, Int)]]
  ): Boolean = {
    val (row, col) = position
    val regionValue = regionGrid.getMatrixValue(position)
    val regionPositions = regionMap(regionValue)

    val adjacentValues = List(
      (row - 1, col),
      (row + 1, col),
      (row, col - 1),
      (row, col + 1)
    ).filter(valueGrid.isValidPosition).map(valueGrid.getMatrixValue)

    val isTopValid = if (valueGrid.isValidPosition((row - 1, col)) && regionGrid.getMatrixValue((row - 1, col)) == regionValue) {
      value < valueGrid.getMatrixValue((row - 1, col))
    } else true

    val isBottomValid = if (valueGrid.isValidPosition((row + 1, col)) && regionGrid.getMatrixValue((row + 1, col)) == regionValue) {
      value > valueGrid.getMatrixValue((row + 1, col))
    } else true

    !(
      valueGrid.getMatrixValue(position) != 0 ||
      value < 1 || value > regionPositions.length ||
      adjacentValues.contains(value) ||
      regionPositions.map(valueGrid.getMatrixValue).contains(value) ||
      !isTopValid ||
      !isBottomValid
    )
  }

  /**
   * Tenta colocar valores válidos na posição atual.
   * Input: Recebe a matriz de valores, de regiões, a linha e coluna atual e uma lista de valores a serem testados.
   * Output:
   * Caso a lista de valores a serem testados esteja vazia, retorna None.
   * Caso o valor possa ser inserido na posição atual, retorna a matriz de valores com o valor inserido.
   * Caso o valor não possa ser inserido na posição atual, chama a função recursivamente com a lista de valores
   * restantes.
   */
  @tailrec
  private def tryValues(
    row: Int,
    col: Int,
    valuesToTry: List[Int],
    valueGrid: Matrix[Int],
    regionGrid: Matrix[String],
    regionMap: mutable.HashMap[String, List[(Int, Int)]]
  ): Option[Matrix[Int]] = {
    if (valuesToTry.isEmpty) {
      None
    } else {
      val value = valuesToTry.head
      if (canInsertValue((row, col), value, valueGrid, regionGrid, regionMap)) {
        val updatedValueGrid = valueGrid.setMatrixValue((row, col), value)
        val nextPosition = if (col == valueGrid.numCols - 1) (row + 1, 0) else (row, col + 1)
        solveKojun(updatedValueGrid, regionGrid, nextPosition, regionMap) match {
          case Some(solution) => Some(solution)
          case None => tryValues(row, col, valuesToTry.tail, valueGrid, regionGrid, regionMap)
        }
      } else {
        tryValues(row, col, valuesToTry.tail, valueGrid, regionGrid, regionMap)
      }
    }
  }

  /**
   * Resolve o Kojun a partir da uma matriz de valores (Int) e da matriz de regiões (Char).
   * Por ser uma função recursiva, recebe um ponto de partida (linha e coluna) para aplicar o backtracking. A função
   * normalmente é chamada externamente com (0, 0).
   * O retorno é um Option que pode ser None caso não exista solução ou Some caso exista.
   */
  @tailrec
  private def solveKojun(
    valueGrid: Matrix[Int],
    regionGrid: Matrix[String],
    position: Position,
    regionMap: mutable.HashMap[String, List[(Int, Int)]]
  ): Option[Matrix[Int]] = {
    val (row, col) = position
    if (row == valueGrid.numRows) {
      Some(valueGrid)
    } else if (col == valueGrid.numCols) {
      solveKojun(valueGrid, regionGrid, (row + 1, 0), regionMap)
    } else if (valueGrid.isValidPosition(position) && valueGrid.getMatrixValue(position) != 0) {
      solveKojun(valueGrid, regionGrid, (row, col + 1), regionMap)
    } else {
      val maxRegionSize = regionMap(regionGrid.getMatrixValue(position)).length
      tryValues(row, col, (1 to maxRegionSize).toList, valueGrid, regionGrid, regionMap)
    }
  }

  def solve(): Option[Matrix[Int]] = {
    /*
    // Testa os diferentes casos de canInsertValue
    println("Can insert value: (0,0), 1 (should be false)")
    println(canInsertValue((0, 0), 1, valueGrid, regionGrid, regionMap))
    println("Can insert value: (5,5), 3 (should be false)")
    println(canInsertValue((5, 5), 3, valueGrid, regionGrid, regionMap))
    println("Can insert value: (5,5), 1 (should be true)")
    println(canInsertValue((5, 5), 1, valueGrid, regionGrid, regionMap))
    println("Can insert value: (1,4), 3 (should be false)")
    println(canInsertValue((1, 4), 3, valueGrid, regionGrid, regionMap))
    println("Can insert value: (1,4), 4 (should be false)")
    println(canInsertValue((1, 4), 4, valueGrid, regionGrid, regionMap))
    println("Can insert value: (1,4), 6 (should be true)")
    println(canInsertValue((1, 4), 6, valueGrid, regionGrid, regionMap))
    println("Can insert value: (1,0), 2 (should be false)")
    println(canInsertValue((1, 0), 2, valueGrid, regionGrid, regionMap))
    println("Can insert value: (1,0), 3 (should be false)")
    println(canInsertValue((1, 0), 3, valueGrid, regionGrid, regionMap))
    */

    solveKojun(valueGrid, regionGrid, (0, 0), regionMap)
  }
}
